lapis/storage.py

```diff
@@ -0,0 +1,81 @@
+from usim import time
+
+from typing import Optional
+
+from lapis.utilities.cache_algorithm_implementations import cache_algorithm
+from lapis.utilities.cache_cleanup_implementations import cache_cleanup
+
+
+class Storage(object):
+
+    __slots__ = ("name", "sitename", "storagesize", "usedstorage", "content")
+
+    def __init__(
+        self, name: str, sitename: str, storagesize: int, content: Optional[dict] = None
+    ):
+        self.name = name
+        self.sitename = sitename
+        self.storagesize = storagesize
+        self.content = content
+        self.usedstorage = self.get_used_storage()
+        self.describe_state()
+
+    def get_used_storage(self):
+        return sum(subdict["usedsize"] for subdict in self.content.values())
+
+    def free_space(self):
+        return self.storagesize - self.usedstorage
+
+    def place_new_file(self, filerequest: tuple):
+        filename, filespecs = filerequest
+        if self.free_space() - filespecs["usedsize"] < 0:
+            self.make_room(self.free_space() - filespecs["usedsize"])
+        self.content.update({filename: filespecs})
+        self.content[filename].update(
+            cachedsince=time.now, lastaccessed=time.now, numberofaccesses=0
+        )
+        self.usedstorage = self.get_used_storage()
+
+    def update_file(self, filerequest: tuple):
+        filename, filespecs = filerequest
+        requested_file = filename
+        filesize_difference = (
+            filespecs["usedsize"] - self.content[requested_file]["usedsize"]
+        )
+        if filesize_difference > 0:
+            self.make_room(filesize_difference)
+            self.content[requested_file]["usedsize"] += filesize_difference
+        self.content[requested_file]["lastaccessed"] = time.now
+        self.content[requested_file]["numberofaccesses"] += 1
+        self.usedstorage = self.get_used_storage()
+
+    def make_room(self, filesize_difference: int):
+        if self.free_space() - filesize_difference < 0:
+            cache_cleanup["fifo"](filesize_difference, self)
+
+    def provides_file(self, filerequest: dict):
+        filename, filespecs = filerequest
+        if filename in self.content.keys():
+            self.update_file(filerequest)
+            return True
+        else:
+            if self.cache_file():
+                self.place_new_file(filerequest)
+            return False
```

Comment:
I am a bit puzzled now. If I call a method named `provides_file` I wouldn't expect it to do an update on the contents of the storage. Further, I would expect a method `storage[filename]` to handle my request. Also for setting, I would expect something like `storage[filename] = meta`. What do you think?

Reply:
Mh, `providing_file` would probably have been a better way of communication what I want this function to do. `storage.provides_file`  is supposed to model the what happens when a file is provided including the changes made to the storage element. 
Would this be more clear?
> Also for setting, I would expect something like storage[filename] = meta

Isn't this what is happening in `update_file/place_new_file`?

Reply:
> > Also for setting, I would expect something like storage[filename] = meta
> 
> Isn't this what is happening in `update_file/place_new_file`?

I mean as interface of the storage itself :) There are some nice methods (``__getitem__`` and ``__setitem__``) that you might implement to make the storage look like a dictionary for example. In case you want to consider this, just ask :)